man/man1/xargs.1
.\" Copyright (c) 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" John B. Roll Jr. and the Institute of Electrical and Electronics
.\" Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)xargs.1	5.5 (Berkeley) 6/27/91
.\"
.TH XARGS 1 "June 27, 1991"
.UC 7
.SH NAME
xargs \- construct argument list(s) and execute utility.
.SH SYNOPSIS
.B xargs
.RB [ \-ft0 ]
.RB [[ \-x ]
.B \-n
.IR number ]
.RB [ \-s
.IR size ]
.RI [ utility
.RI [ argument " ...]]"
.SH DESCRIPTION
The
.B xargs
utility reads space, tab, newline and end-of-file delimited arguments
from the standard input and executes the specified
.I utility
with them as arguments.
.PP
The utility and any arguments specified on the command line are given
to the
.I utility
upon each invocation, followed by some number of the arguments read
from standard input.
The
.I utility
is repeatedly executed until standard input is exhausted.
.PP
Spaces, tabs and newlines may be embedded in arguments using single (`` ' '')
or double (``"'') quotes or backslashes (``\e'').
Single quotes escape all non-single quote characters, excluding newlines,
up to the matching single quote.
Double quotes escape all non-double quote characters, excluding newlines,
up to the matching double quote.
Any single character, including newlines, may be escaped by a backslash.
.PP
The options are as follows:
.TP
.BI \-n " number"
Set the maximum number of arguments taken from standard input for each
invocation of the utility.
An invocation of
.I utility
will use less than
.I number
standard input arguments if the number of bytes accumulated (see the
.I \-s
option) exceeds the specified
.I size
or there are fewer than
.I number
arguments remaining for the last invocation of
.IR utility .
The current default value for
.I number
is 5000.
.TP
.BI \-s " size"
Set the maximum number of bytes for the command line length provided to
.IR utility .
The sum of the length of the utility name and the arguments passed to
.I utility
(including NULL terminators) will be less than or equal to this number.
The current default value for
.I size
is ARG_MAX - 2048.
.TP
.B \-t
Echo the command to be executed to standard error immediately before it
is executed.
.TP
.B \-x
Force
.B xargs
to terminate immediately if a command line containing
.I number
arguments will not fit in the specified (or default) command line length.
.TP
.B \-0
Read null-byte terminated pathnames from standard input as may have been
produced by the
.B \-print0
option of
.BR find (1).
This is a MINIX 3 specific extension to
.BR xargs .
.PP
If no
.I utility
is specified,
.BR echo (1)
is used.
.PP
Undefined behavior may occur if
.I utility
reads from the standard input.
.PP
.B Xargs
exits with an exit status of 0 if no error occurs.
If
.I utility
cannot be invoked, is terminated by a signal or terminates without
calling
.BR exit (2),
.B xargs
exits with an exit status of 127.
If
.I utility
exits with an exit status other than 0,
.B xargs
exits with that exit status.
Otherwise,
.B xargs
exits with an exit status of 1.
.SH "SEE ALSO"
.BR echo (1),
.BR find (1).
.SH STANDARDS
The
.B xargs
utility is expected to be POSIX 1003.2 compliant.


commands/xargs/xargs.c
/*-
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * John B. Roll Jr.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1990 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)xargs.c	5.11 (Berkeley) 6/19/91";
#endif /* not lint */

#include <sys/types.h>
#include <sys/wait.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>
#include <fcntl.h>
#include <stdarg.h>
#if __minix
#define _PATH_ECHO	"/bin/echo"
#else
#include "pathnames.h"
#endif

#ifndef ARG_MAX
#define ARG_MAX		(sizeof(int) == 2 ? 4096 : 128 * 1024)
#endif

int exit_status = 0;
int tflag;
void err(const char *, ...);
void run(char **argv);
void usage(void);

int main(int argc, char **argv)
{
	extern int optind;
	extern char *optarg;
	register int ch;
	register char *p, *bbp, *ebp, **bxp, **exp, **xp;
	int cnt, indouble, insingle, nargs, nflag, nline, xflag, zflag;
	char **av, *argp;

	/*
	 * POSIX.2 limits the exec line length to ARG_MAX - 2K.  Running that
	 * caused some E2BIG errors, so it was changed to ARG_MAX - 4K.  Given
	 * that the smallest argument is 2 bytes in length, this means that
	 * the number of arguments is limited to:
	 *
	 *	 (ARG_MAX - 4K - LENGTH(utility + arguments)) / 2.
	 *
	 * We arbitrarily limit the number of arguments to 5000.  This is
	 * allowed by POSIX.2 as long as the resulting minimum exec line is
	 * at least LINE_MAX.  Realloc'ing as necessary is possible, but
	 * probably not worthwhile.
	 */
#if !__minix || __minix_vmd
	nargs = 5000;
	nline = ARG_MAX - 4 * 1024;
#else
	/* Things are more cramped under standard Minix. */
	nargs = 80 * sizeof(int);
	nline = ARG_MAX - 512 * sizeof(int);
#endif
	nflag = xflag = zflag = 0;
	while ((ch = getopt(argc, argv, "n:s:tx0")) != EOF)
		switch(ch) {
		case 'n':
			nflag = 1;
			if ((nargs = atoi(optarg)) <= 0)
				err("illegal argument count");
			break;
		case 's':
			nline = atoi(optarg);
			break;
		case 't':
			tflag = 1;
			break;
		case 'x':
			xflag = 1;
			break;
		case '0':
			zflag = 1;
			break;
		case '?':
		default:
			usage();
	}
	argc -= optind;
	argv += optind;

	if (xflag && !nflag)
		usage();

	/*
	 * Allocate pointers for the utility name, the utility arguments,
	 * the maximum arguments to be read from stdin and the trailing
	 * NULL.
	 */
	if (!(av = bxp =
	    malloc((u_int)(1 + argc + nargs + 1) * sizeof(char **))))
		err("%s", strerror(errno));

	/*
	 * Use the user's name for the utility as argv[0], just like the
	 * shell.  Echo is the default.  Set up pointers for the user's
	 * arguments.
	 */
	if (!*argv)
		cnt = strlen(*bxp++ = _PATH_ECHO);
	else {
		cnt = 0;
		do {
			cnt += strlen(*bxp++ = *argv) + 1;
		} while (*++argv);
	}

	/*
	 * Set up begin/end/traversing pointers into the array.  The -n
	 * count doesn't include the trailing NULL pointer, so the malloc
	 * added in an extra slot.
	 */
	exp = (xp = bxp) + nargs;

	/*
	 * Allocate buffer space for the arguments read from stdin and the
	 * trailing NULL.  Buffer space is defined as the default or specified
	 * space, minus the length of the utility name and arguments.  Set up
	 * begin/end/traversing pointers into the array.  The -s count does
	 * include the trailing NULL, so the malloc didn't add in an extra
	 * slot.
	 */
	nline -= cnt;
	if (nline <= 0)
		err("insufficient space for command");

	if (!(bbp = malloc((u_int)nline + 1)))
		err("%s", strerror(errno));
	ebp = (argp = p = bbp) + nline - 1;

	if (zflag) {
		/* Read pathnames terminated by null bytes as produced by
		 * find ... -print0.  No comments in this code, see further
		 * below.
		 */
		for (;;)
			switch(ch = getchar()) {
			case EOF:
				if (p == bbp)
					exit(exit_status);

				if (argp == p) {
					*xp = NULL;
					run(av);
					exit(exit_status);
				}
				/*FALL THROUGH*/
			case '\0':
				if (argp == p)
					continue;

				*p = '\0';
				*xp++ = argp;

				if (xp == exp || p == ebp || ch == EOF) {
					if (xflag && xp != exp && p == ebp)
						err(
					   "insufficient space for arguments");
					*xp = NULL;
					run(av);
					if (ch == EOF)
						exit(exit_status);
					p = bbp;
					xp = bxp;
				} else
					++p;
				argp = p;
				break;
			default:
				if (p < ebp) {
					*p++ = ch;
					break;
				}

				if (bxp == xp)
					err("insufficient space for argument");
				if (xflag)
					err("insufficient space for arguments");

				*xp = NULL;
				run(av);
				xp = bxp;
				cnt = ebp - argp;
				bcopy(argp, bbp, cnt);
				p = (argp = bbp) + cnt;
				*p++ = ch;
				break;
			}
		/* NOTREACHED */
	}

	for (insingle = indouble = 0;;)
		switch(ch = getchar()) {
		case EOF:
			/* No arguments since last exec. */
			if (p == bbp)
				exit(exit_status);

			/* Nothing since end of last argument. */
			if (argp == p) {
				*xp = NULL;
				run(av);
				exit(exit_status);
			}
			goto arg1;
		case ' ':
		case '\t':
			/* Quotes escape tabs and spaces. */
			if (insingle || indouble)
				goto addch;
			goto arg2;
		case '\n':
			/* Empty lines are skipped. */
			if (argp == p)
				continue;

			/* Quotes do not escape newlines. */
arg1:			if (insingle || indouble)
				 err("unterminated quote");

arg2:			*p = '\0';
			*xp++ = argp;

			/*
			 * If max'd out on args or buffer, or reached EOF,
			 * run the command.  If xflag and max'd out on buffer
			 * but not on args, object.
			 */
			if (xp == exp || p == ebp || ch == EOF) {
				if (xflag && xp != exp && p == ebp)
					err("insufficient space for arguments");
				*xp = NULL;
				run(av);
				if (ch == EOF)
					exit(exit_status);
				p = bbp;
				xp = bxp;
			} else
				++p;
			argp = p;
			break;
		case '\'':
			if (indouble)
				goto addch;
			insingle = !insingle;
			break;
		case '"':
			if (insingle)
				goto addch;
			indouble = !indouble;
			break;
		case '\\':
			/* Backslash escapes anything, is escaped by quotes. */
			if (!insingle && !indouble && (ch = getchar()) == EOF)
				err("backslash at EOF");
			/* FALLTHROUGH */
		default:
addch:			if (p < ebp) {
				*p++ = ch;
				break;
			}

			/* If only one argument, not enough buffer space. */
			if (bxp == xp)
				err("insufficient space for argument");
			/* Didn't hit argument limit, so if xflag object. */
			if (xflag)
				err("insufficient space for arguments");

			*xp = NULL;
			run(av);
			xp = bxp;
			cnt = ebp - argp;
			bcopy(argp, bbp, cnt);
			p = (argp = bbp) + cnt;
			*p++ = ch;
			break;
		}
	/* NOTREACHED */
}

void run(char **argv)
{
	register char **p;
	pid_t pid;
	int noinvoke;
	int status;
	int pfd[2];

	if (tflag) {
		(void)fprintf(stderr, "%s", *argv);
		for (p = argv + 1; *p; ++p)
			(void)fprintf(stderr, " %s", *p);
		(void)fprintf(stderr, "\n");
		(void)fflush(stderr);
	}
	if (pipe(pfd) < 0) err("pipe: %s", strerror(errno));

	switch(pid = fork()) {
	case -1:
		err("fork: %s", strerror(errno));
	case 0:
		close(pfd[0]);
		fcntl(pfd[1], F_SETFD, fcntl(pfd[1], F_GETFD) | FD_CLOEXEC);

		execvp(argv[0], argv);
		noinvoke = (errno == ENOENT) ? 127 : 126;
		(void)fprintf(stderr,
		    "xargs: %s exec failed: %s.\n", argv[0], strerror(errno));

		/* Modern way of returning noinvoke instead of a dirty vfork()
		 * trick:					(kjb)
		 */
		write(pfd[1], &noinvoke, sizeof(noinvoke));
		_exit(-1);
	}
	close(pfd[1]);
	if (read(pfd[0], &noinvoke, sizeof(noinvoke)) < sizeof(noinvoke))
		noinvoke = 0;
	close(pfd[0]);

	pid = waitpid(pid, &status, 0);
	if (pid == -1)
		err("waitpid: %s", strerror(errno));

	/*
	 * If we couldn't invoke the utility or the utility didn't exit
	 * properly, quit with 127 or 126 respectively.
	 */
	if (noinvoke)
		exit(noinvoke);

	/*
	 * According to POSIX, we have to exit if the utility exits with
	 * a 255 status, or is interrupted by a signal.   xargs is allowed
	 * to return any exit status between 1 and 125 in these cases, but
	 * we'll use 124 and 125, the same values used by GNU xargs.
	 */
	if (WIFEXITED(status)) {
		if (WEXITSTATUS (status) == 255) {
			fprintf (stderr, "xargs: %s exited with status 255\n",
				 argv[0]);
			exit(124);
		} else if (WEXITSTATUS (status) != 0) {
			exit_status = 123;
		}
	} else if (WIFSTOPPED (status)) {
		fprintf (stderr, "xargs: %s terminated by signal %d\n",
			 argv[0], WSTOPSIG (status));
		exit(125);
	} else if (WIFSIGNALED (status)) {
		fprintf (stderr, "xargs: %s terminated by signal %d\n",
			 argv[0], WTERMSIG (status));
		exit(125);
	}
}

void usage(void)
{
	(void)fprintf(stderr,
"usage: xargs [-t0] [[-x] -n number] [-s size] [utility [argument ...]]\n");
	exit(1);
}

void err(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	(void)fprintf(stderr, "xargs: ");
	(void)vfprintf(stderr, fmt, ap);
	va_end(ap);
	(void)fprintf(stderr, "\n");
	exit(1);
	/* NOTREACHED */
}
